File: lib-google-proximity/src/main/java/com/ndipatri/iot/googleproximity/models/api/Beacon.java
package com.ndipatri.iot.googleproximity.models.api;

import android.util.Base64;

import com.ndipatri.iot.googleproximity.utils.ByteUtils;

public class Beacon {

    // These constants are in the Proximity Service Status enum:
    public static final String STATUS_NOT_FOUND = "NOT_FOUND";
    public static final String STATUS_UNREGISTERED = "UNREGISTERED";
    public static final String STATUS_ACTIVE = "ACTIVE";
    public static final String STATUS_DECOMMISSIONED = "DECOMMISSIONED";
    public static final String STATUS_INACTIVE = "INACTIVE";

    public AdvertisedId advertisedId;

    public String status;
    public String placeId;

    public Beacon(byte[] advertiseId) {
        this("EDDYSTONE", advertiseId, Beacon.STATUS_ACTIVE);
    }

    public Beacon(String type, byte[] rawId, String status) {
        this.advertisedId = new AdvertisedId(type, rawId);
        this.status = status;
        this.placeId = null;
    }

    /**
     * The beaconName is formatted as "beacons/%d!%s" where %d is an integer representing the
     * beacon ID type. For Eddystone this is 3. The %s is the base16 (hex) ASCII for the ID bytes.
     */
    public String getBeaconName() {

        byte[] rawId = Base64.decode(advertisedId.id, Base64.DEFAULT);
        String hexEncodedId = ByteUtils.convertBytesToHex(rawId);

        return "3!" + hexEncodedId;
    }
}

File: lib-google-proximity/src/main/java/com/ndipatri/iot/googleproximity/models/api/Beacon.kt
package com.ndipatri.iot.googleproximity.models.api

import android.util.Base64
import kotlin.experimental.and

class Beacon(type: String, rawId: ByteArray, var status: String) {

    var advertisedId: AdvertisedId
    var placeId: String? = null

    /**
     * The beaconName is formatted as "beacons/%d!%s" where %d is an integer representing the
     * beacon ID type. For Eddystone this is 3. The %s is the base16 (hex) ASCII for the ID bytes.
     */
    val beaconName: String
        get() {

            val rawId = Base64.decode(advertisedId.id, Base64.DEFAULT)
            val hexEncodedId = convertBytesToHex(rawId)

            return "3!$hexEncodedId"
        }

    constructor(advertiseId: ByteArray) : this("EDDYSTONE", advertiseId, Beacon.STATUS_ACTIVE) {}

    init {
        this.advertisedId = AdvertisedId(type, rawId)
        this.placeId = null
    }

    private val HEX_ARRAY = "0123456789ABCDEF".toCharArray()
    @ExperimentalUnsignedTypes
    fun convertBytesToHex(bytes: ByteArray): String {
        val hex = CharArray(bytes.size * 2)
        for (i in bytes.indices) {
            val v = bytes[i].toUInt() and 0xFF.toUInt()
            hex[i * 2] = HEX_ARRAY[v.toInt() ushr 4]
            hex[i * 2 + 1] = HEX_ARRAY[(v and 0x0F.toUInt()).toInt()]
        }

        return String(hex).toLowerCase()
    }

    companion object {

        // These constants are in the Proximity Service Status enum:
        val STATUS_NOT_FOUND = "NOT_FOUND"
        val STATUS_UNREGISTERED = "UNREGISTERED"
        val STATUS_ACTIVE = "ACTIVE"
        val STATUS_DECOMMISSIONED = "DECOMMISSIONED"
        val STATUS_INACTIVE = "INACTIVE"
    }
}
